### ISSUE TITLE
Product Block Editor: Add drag and drop support to Attributes block

### ISSUE THREAD

**Issue by louwie17** [woothemes/woocommerce#38189]
**Description**

Users should be able to re-arrange the order of attributes. We are allowing this by the use of drag and drop.
We should add drag and drop support to attributes, this is currently only possible by turning each attribute within the list into its own block.
This would be similar to the List block, where an item within the list can be dragged.

https://user-images.githubusercontent.com/2240960/237085336-5bfa99c5-ba97-4fa0-9820-48c5363ccc88.mp4

**Acceptance criteria**

- [ ] Attributes can be re-ordered by the use of drag and drop

**Comment by joshuatf** [comment]
@jarekmorawski Do we want to reimplement the drag handle where it was on the left-hand side of each attribute?  Or do we want to add a toolbar with a drag button like Gutenberg does for its blocks?

Dev notes: we can find the drag handle here - https://github.com/WordPress/gutenberg/blob/b2c16f3c684098a8dabff3911c610caa5c053701/packages/block-editor/src/components/block-mover/index.js#L67-L80

We should also constrain the new attribute blocks to only the parent attributes block so they cannot be dragged outside of it.

**Comment by jarekmorawski** [comment]
It's a tough question. I discussed this with @pmcpinto the other day, and we've decided to keep the existing layout for now. It's possible that we'll migrate to contextual controls in the toolbar in the future, but given GB's current adoption, that interaction model is foreign to most Woo merchants and may hurt discoverability.

**Comment by pmcpinto** [comment]
@louwie17 we should do a spike on this in cycle 43.

**Comment by louwie17** [comment]
> @louwie17 we should do a spike on this in cycle 43.

Yeah that be good, this was work that got started by Matt, but we end up pushing it out because of other priorities.

**Comment by j111q** [comment]
Thanks for flagging this Matt! 
I recently examined Dragging in the Attributes table and these are my thoughts:

1. Drag-to-sort is an available feature in the legacy editor, but not currently in our new Product Form. From my tests, I believe the Attribute Sorting order would **only impact how the Attribute is displayed in the Storefront**, under Description.
2. In the legacy editor, it also impacts the order of the Variation options, but we’ve already moved that to Variations. So if we decide to allow the sorting functionality, the merchant should be sorting the rows under the Variations tab (although we don’t allow it right now in the new Product Form). --> ie. Drag-to-sort for the Variations table would be separate issue, as you've created here ( #45775 )
3. For the reasons above, only Attributes that are marked as “Visible on the product page” should be at the top of the column and should be draggable.
4. All other attributes should **not** be draggable, and they could be arranged alphabetically under the draggable attributes.

This is a mock-up to show what I mean:
![image](https://github.com/woocommerce/woocommerce/assets/65582817/ee21f0a0-fdb8-4df8-ac5c-4622a9303f70)

Does this make sense? ☝️ 

For more context, you can also go to this P2 post: peCask-GW-p2

**Comment by pmcpinto** [comment]
@j111q in this epic https://github.com/woocommerce/woocommerce/issues/36286, we defined the following logic to sort attributes and variation options:

> Attributes and Variation options can be sorted separately. Variation options are, by default, hidden from the product description. If the user enables the visibility toggle, Variation options will be displayed below regular attributes.
> Changing the options order will be reflected at the top of the page and in the product description.
> If a product has 7 attributes: 2 variation options and 5 regular attributes, they can move n2 above n1, but not n7 above n2. They are managed in separate lists, even though they’re displayed in the same area in the product description.

Is this aligned with what you have in mind? 

> In the legacy editor, it also impacts the order of the Variation options, but we’ve already moved that to Variations

Do you mean the sort order feature available in the Global Attributes management flow outside the product form?

**Comment by pmcpinto** [comment]
@j111q , can you check if the logic shared above aligns with your proposal? Thanks

**Comment by j111q** [comment]
> Is this aligned with what you have in mind?
I actually was thinking it should work like this:

1. Creating a variation option automatically generates the attribute in the attributes table. (This is something we discussed based on user feedback in the interviews, plus, the need for variation options to be mapped as attributes in multichannel.)
2. `Show in product details` no longer appears under Variation options -- Instead, it appears under the automatically generated `Attribute`. 
3. From the `Attributes` table, users can then drag and sort all the attributes which have `Show in product details` checked. 

Does that make sense?

> Do you mean the sort order feature available in the Global Attributes management flow outside the product form?

No, I meant that currently in the legacy editor, if you sort the Attributes in a certain way, (eg. Red, Yellow, Blue, Green), it will be displayed in the dropdown of the product as Red, Yellow, Blue, Green. This is true **even when you have ordered your variations in a different way**. So if your variations are ordered Blue, Yellow, Red, Green, but your attributes are Red, Yellow, Blue, Green, it will be displayed in the dropdown as Red, Yellow, Blue, Green. Here's a [screencast](https://www.dropbox.com/scl/fi/w6jwa6iqxa32agfamasvs/variation-order.gif?rlkey=677hgvdww9aotx5kpsnkzq18c&dl=0) of me testing this. 

My point was that, the order of variation options should be determined by the order set under Variations, **not by the order of Attributes**.

**Comment by pmcpinto** [comment]
@j111q thanks for the clarification. Why would the user want to reorder individual variations? Eg: Blue + XL.

The variations order in the front-end should be influenced by two levels:
- The variation options (Attributes) order. Eg: Color and Size.
- The variation option values (Attribute values) order. Eg: Green and Red. 

Does this make sense?

**Comment by ResponsiveFlow** [comment]
@pmcpinto 

> Why would the user want to reorder individual variations?

Because if I add size as a variation and I add the options: S, M, L, XL
They're then reordered alphabetically as L, M, S, XL which makes no sense.

**Comment by pmcpinto** [comment]
I'm going to close this one, as it refers to the new editor who will be removed.
